Add ScrollView.automaticallyAdjustsScrollIndicatorInsets prop (on iOS)

justinwh opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary

iOS 13 added a new property to `UIScrollView`: `automaticallyAdjustsScrollIndicatorInsets`, which is `YES` by default.  The property changes the meaning of the `scrollIndicatorInsets` property.  When `YES`, any such insets are **in addition to** whatever insets would be applied by the device's safe area.  When `NO`, the iOS <13 behavior is restored, which is for such insets to not account for safe area.

In other words, this effects ScrollViews that underlay the device's safe area (i.e. under the notch).  When `YES`, the OS "automatically" insets the scroll indicators, when `NO` it does not.

There are two problems with the default `YES` setting:

1. It means applying `scrollIndicatorInsets` to a `ScrollView` has a different effect on iOS 13 versus iOS 12.
2. It limits developers' control over `scrollIndicatorInsets`.  Since negative insets are not supported, if the insets the OS chooses are too large for your app, you cannot fix it.

Further explanation & sample code is available in issue #28140 .

This change sets the default for this property to `NO`, making the behavior consistent across iOS versions, and allowing developers full control.

## Changelog

<!-- Help reviewers and the release process by writing your own changelog entry. For an example, see:
https://github.com/facebook/react-native/wiki/Changelog
-->

[iOS] [Changed] - ScrollView scrollIndicatorInsets to not automatically add safe area on iOS13+

## Test Plan

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

Here are screenshots of the demo app (from the original bug) before (with safe area applied to insets) & after (without safe area applied to insets):

![before](https://user-images.githubusercontent.com/428831/91644197-ea03a700-ea07-11ea-9489-be27820930eb.png)

![after](https://user-images.githubusercontent.com/428831/91644200-eff98800-ea07-11ea-8788-daf1e783639d.png)

justinwh commented:
There is an extra concern with this PR: changing the default behavior will require developers who are using ScrollView in this way to re-adjust their code.

- for devs who have already added their own workarounds, they'll need to revert them

- for devs that depended on the previous iOS 13 default behavior (and aren't concerned about iOS <=12), they'll need to explicitly add appropriate insets

Based on the low interest in the initial bug, I don't think this is a lot of people, but it's not zero either.

osartun commented:
What are the next steps for this PR? We have a local patch for RN in our repo with pretty much the same code. I would like to see that issue fixed so that we don't need to reapply our patch with every upgrade of RN.

> There is an extra concern with this PR: changing the default behavior will require developers who are using ScrollView in this way to re-adjust their code.
> 
> * for devs who have already added their own workarounds, they'll need to revert them
> * for devs that depended on the previous iOS 13 default behavior (and aren't concerned about iOS <=12), they'll need to explicitly add appropriate insets

Can't you just put this into a new "breaking" version? Whenever we upgrade we check the changes and are expecting something to break.
Seeing this getting fixed would be a satisfying breaking change though.

jfrolich commented:
I'd love that someone in the core team would look at this. Without this, it is impossible to get the scroll indicators right on iOS in some scenarios. It's also a bit frustrating because I think a lot of people would fiddle with it and conclude it's not possible to get it right, or after a long time find this patch (like me 😃)

osartun commented:
There are also multiple issues and implemented workarounds associated to this which would get closed:

* https://github.com/facebook/react-native/issues/26610
* https://github.com/facebook/react-native/issues/28140
* https://github.com/facebook/react-native/issues/28085
* https://github.com/standardnotes/mobile/issues/192
* https://github.com/mochixuan/react-native-drag-sort/pull/92
* https://github.com/shoutem/ui/pull/493
* …

justinwh commented:
Yeah, crazy huh?  I'm the original submitter—I don't know what else to do to get this fixed.  I documented the issue pretty thoroughly, found the cause, and submitted a fix PR.  I'm not authorized to merge it myself.

As far as I can tell, nobody from the project who is authorized to merge it is aware it exists.

I don't know what to do to get a maintainer's attention.  Today I tried asking in the Reactiflux discord linked from the project website, but didn't get any useful answers there either.

The original issue (which is now just shy of its one-year anniversary) still displays the "Needs: Triage" label...

On the plus side, I can confirm that I've been shipping this fix in my own app for over a year and it has not caused any problems.

lunaleaps commented:
These changes look good to me! cc @PeteTheHeat or @p-sun for any thoughts?

lunaleaps commented:
@justinwh Would it be possible to add your test case (demonstrating the scrollbar positions) as an RNTester case? How were you able to display the scrollbars persistently? We're trying to see how we can prevent this from regressing in the future.

justinwh commented:
> @justinwh Would it be possible to add your test case (demonstrating the scrollbar positions) as an RNTester case? How were you able to display the scrollbars persistently? We're trying to see how we can prevent this from regressing in the future.

@lunaleaps Unfortunately, I captured my screenshots manually without persistent scrollbars (i.e. just kept trying until I got the timing right).

As far as I know the scrollbars cannot be made to display persistently on iOS.  And I think their animation is too quick to capture consistently in a screenshot-based automated test.

I'm not aware of any good option for setting up a test case for this.  (But I'm open to suggestions.)

lunaleaps commented:
> As far as I know the scrollbars cannot be made to display persistently on iOS. And I think their animation is too quick to capture consistently in a screenshot-based automated test.

Yeah, I agree. I think then a manual test example case of this might be the best we can do? cc @p-sun who had some thoughts on the changes.

justinwh commented:
I added a second commit with an rn-tester screen that demonstrates the behavior (called `ScrollViewIndicatorInsets`).

When the fix is applied, this shows the scrollbars passing outside the device safe area to the screen edge (as intended).  If you revert the fix, the scrollbars stay inside the device safe area.

lunaleaps commented:
@facebook-github-bot import

justinwh commented:
I updated the PR with the changes discussed in the issue:
- Made `automaticallyAdjustsScrollIndicatorInsets` a prop on `ScrollView`
- Made the default `true`

I followed the implementation of the `contentInsetAdjustmentBehavior` prop as my template.  What I pushed is sufficient to get the rn-tester example working.  But I also found places in the tree where these props seem to be defined/redefined that I'm less familiar with.  Should I add this new prop into the following files too?

https://github.com/facebook/react-native/blob/d1ab03235cb4b93304150878d2b9057ab45bba77/React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.mm#L257-L270

https://github.com/facebook/react-native/blob/d1ab03235cb4b93304150878d2b9057ab45bba77/ReactCommon/react/renderer/components/scrollview/ScrollViewProps.cpp#L166-L170

lunaleaps commented:
> I updated the PR with the changes discussed in the issue:
> 
> * Made `automaticallyAdjustsScrollIndicatorInsets` a prop on `ScrollView`
> * Made the default `true`
> 
> I followed the implementation of the `contentInsetAdjustmentBehavior` prop as my template. What I pushed is sufficient to get the rn-tester example working. But I also found places in the tree where these props seem to be defined/redefined that I'm less familiar with. Should I add this new prop into the following files too?
> 
> https://github.com/facebook/react-native/blob/d1ab03235cb4b93304150878d2b9057ab45bba77/React/Fabric/Mounting/ComponentViews/ScrollView/RCTScrollViewComponentView.mm#L257-L270
> 
> https://github.com/facebook/react-native/blob/d1ab03235cb4b93304150878d2b9057ab45bba77/ReactCommon/react/renderer/components/scrollview/ScrollViewProps.cpp#L166-L170

Re: ScrollViewProps.cpp, I think this is the defaults we're using for the Fabric native view so we should probably set the default of true there as well.

Re: RCTScrollViewComponentView.mm I believe so yes, as in updating the prop value.

Let me see if someone from with more Fabric experience can confirm. cc @p-sun

p-sun commented:
Yes, @lunaleaps is right. Any new prop should be added to the Fabric component as well. You can add the prop underneath the contentInsetAdjustmentBehavior prop in the two places that Luna linked, add a breakpoint inside RCTScrollViewComponentView.mm, and test it inside the RN Tester app to make sure it works.

Thanks for adding a Switch in your example for this prop. 👍

fkgozali commented:
Btw, RNTester iOS/Android should already support Fabric, but perhaps with some minor quirks. You can enable Fabric in RNTester iOS manually by setting this env var before `pod install`: https://github.com/facebook/react-native/blob/master/packages/rn-tester/Podfile#L22-L30

```
USE_FABRIC=1 pod install
```

Then at least you can verify whether the updated .mm/C++ files compiled and ran fine.

justinwh commented:
I added the Fabric changes, and confirmed they compile and run when `USE_FABRIC=1`.

The rn-tester page I added requires `Modal`, which apparently isn't implemented in Fabric yet.  But I confirmed that setting the prop on other ScrollViews in rn-tester has the correct effect.

lunaleaps commented:
For the example, I noticed that there's no difference between true/false for the property in the case of displaying a Modal with `pageSheet` presentation style. @justinwh was there a reason to include that example? Additionally there's a typo on the `fullScreen` example which I've fixed on the imported PR. I'm going to add some explanation to what we should expect to see when you toggle this on/off. 

Additionally did we discuss adding documentation for this?

justinwh commented:
@lunaleaps thank you for fixing the typo.

The `pageSheet` modal *does* show different behavior for the checked/unchecked states for me.  I'm not sure why yours doesn't, but since this is an "automatic" property I guess nobody but Apple knows how it actually works.  Maybe it's device or OS version dependent?  (I tested on the Simulator with the `iOS 14.5/iPhone 12 Pro` device type and on a real iPhone 12 Mini running iOS 14.6.)

On both devices I tested, enabling it adds a top inset to the `pageSheet` modal, which is incorrect (I think) since the modal does not underlay the statusbar/notch area.  So the behavior you're getting is actually an improvement.

In any case, what it does when enabled is outside our control.  The important thing is that it leaves the scroll indicator insets alone when disabled, and it looks like that's the case in your screenshots.  So I believe this is acceptable.  If you want to remove the `pageSheet` example from rn-tester, I have no objection.

We didn't discuss documentation.  I assumed the docs were generated from the comments, but it sounds like that's not the case.  Is there a process for synchronizing the updates to the react-native-website repo with React Native releases that I should be aware of?

Here's my proposed documentation for the new prop (let me know if you have feedback on it):
```
automaticallyAdjustsScrollIndicatorInsets
Controls whether iOS should automatically adjust the scroll indicator insets. The default value is true. Available on iOS 13 and later.
@platform ios
```

lunaleaps commented:
@justinwh Sounds good! I'll go ahead and land.. 

Regarding docs that sounds good! I wonder if we should also link to the UIScrollView docs for the property since. Would you like to make a PR for react-native-website? It should go under the versioned_docs `next` I believe. Let me know if you need more help!

justinwh commented:
I added a link to the `UIScrollView` docs and sent a PR to the docs repo:

https://github.com/facebook/react-native-website/pull/2661

glenna commented:
hey @justinwh we are integrating this change into our app now and _finally_ removing our RN patch for this, but are confused a bit on what the default actually is. From the code, it looks like it defaults to `true` / `YES` (in iOS world), as well as in the [RN documentation](https://reactnative.dev/docs/scrollview#automaticallyadjustsscrollindicatorinsets-ios), but in this PR description and commit message, it says

> This change sets the default for this property to NO, making the behavior consistent across iOS versions, and allowing developers full control.

What is meant?

justinwh commented:
@glenna The docs are correct—the default is `true`.  I didn't realize the PR's merge commit msg comes from the PR's description (which reflects my original approach, but not the final fix) instead of the commit itself (which I *did* update).

Sorry for the confusion.

glenna commented:
@justinwh not a problem! thanks for the clarification!
